## After removing some words from the csv file, 

_3b) importing necessary libraries and modules._

In [1]:
import turicreate as tc
import utils

_3c) Loading the updated file,_ _**IMDB_Dataset_mod.csv**_ _into an SFrame._ 

_The words that have been removed include:_

has, Has, have, Have, had, Had, do, Do, does, Does, did, Did, is, Is, am, Am, are, Are, was, Was, were, Were, will, Will, shall, Shall, should, Should, can, Can, could, Could, of, Of, for, For, the, The, a, A, an, An, and, And, or, Or, but, But, while, While, whereas, Whereas, I, me, Me, mine, Mine, he, He, his, His, him, Him, she, She, her, Her, hers, Hers, it, It, its, Its, it's, It's, we, We, our, Our, ours, Ours, you, You, yours, Yours, they, They, their, Their, us, Us, only, Only, just, Just, because, Because........ and few more.

_These words and many more which were the neutral words, meaning, unimportant words, were removed. This is because these words lacked sentiment and as these words themselves were unimportant, they were just adding unecessary weights to the weight of the review to predict the overall sentiment. Removing these would help improve the performance of our model and give more accurate predictions._


_**Printing the SFrame**_



In [2]:
movies = tc.SFrame('./IMDB_Dataset_mod.csv')
movies

Finished parsing file /home/hdhesi/SupervisedML/Assignment4/IMDB_Dataset_mod.csv

Parsing completed. Parsed 100 lines in 0.381793 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 49115 lines. Lines per second: 61622.8

Finished parsing file /home/hdhesi/SupervisedML/Assignment4/IMDB_Dataset_mod.csv

Parsing completed. Parsed 50000 lines in 0.810906 secs.

Columns:
	review	str
	sentiment	str

Rows: 50000

Data:
+-------------------------------+-----------+
|             review            | sentiment |
+-------------------------------+-----------+
| One reviewers mentioned af... |  positive |
|  wonderful little producti... |  positive |
|  thought this wonderful wa... |  positive |
| Basically there's family w... |  negative |
| Petter Mattei's "Love in T... |  positive |
| Probably my all-time favor... |  positive |
|  sure like to see resurrec... |  positive |
| This show amazing, fresh &... |  negative |
| Encouraged by positive com... |  negative |
| If like original gut wrenc... |  positive |
+-------------------------------+-----------+
[50000 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

_3d) Adding a column ' words' to the SFrame to store the count of each word used in each review._

_**Printing the SFrame**_

In [3]:
movies['words'] = tc.text_analytics.count_words(movies['review'])
movies

Columns:
	review	str
	sentiment	str
	words	dict

Rows: 50000

Data:
+-------------------------------+-----------+-------------------------------+
|             review            | sentiment |             words             |
+-------------------------------+-----------+-------------------------------+
| One reviewers mentioned af... |  positive | {'darker': 1.0, 'touch': 1... |
|  wonderful little producti... |  positive | {'terribly': 1.0, 'surface... |
|  thought this wonderful wa... |  positive | {'go': 1.0, 'great': 1.0, ... |
| Basically there's family w... |  negative | {'them': 1.0, 'ignore': 1.... |
| Petter Mattei's "Love in T... |  positive | {'work': 1.0, 'anxiously':... |
| Probably my all-time favor... |  positive | {'dozen': 1.0, 'if': 1.0, ... |
|  sure like to see resurrec... |  positive | {'it': 1.0, 'then': 1.0, '... |
| This show amazing, fresh &... |  negative | {'awful': 1.0, 'now': 1.0,... |
| Encouraged by positive com... |  negative | {'obsessives': 1.0, 'effor... |
| If like original gut wrenc... |  positive | {'mom': 1.0, 'my': 1.0, 'h... |
+-------------------------------+-----------+-------------------------------+
[50000 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

_3e) Randomly splitting the data into training and testing data, with 80% data in the training set and remaining 20% in the testing set._

In [4]:
train, test = movies.random_split(.8, seed=0)

_3f) Creating a logistic Classifier that uses both L1 and L2 penalty of my choice (L1 penalty = 0.0, L2 penalty = 0.01) and training it on training data._

In [5]:
model = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty = 0.0, l2_penalty = 0.01)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92600

Number of coefficients      : 92601

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 1.117250     | 0.926085          | 0.855145            |

| 1         | 4        | 1.000000  | 1.256063     | 0.970813          | 0.865135            |

| 2         | 5        | 1.000000  | 1.343426     | 0.980936          | 0.889111            |

| 3         | 7        | 0.975237  | 1.478998     | 0.986984          | 0.892607            |

| 4         | 8        | 1.000000  | 1.572172     | 0.992217          | 0.895604            |

| 9         | 13       | 1.000000  | 2.034029     | 0.999448          | 0.868132            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

_The_ _**Training accuracy**_ _comes out to be_ _**0.999448 (99.94%)**_ _and the_ _**Validation accuracy**_ _is_ _**0.868132 (86.81%)**_.     

_**Printing the model**_

In [6]:
model

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 92601
Number of examples             : 38030
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 92600

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 10
Solver status                  : Completed (Iteration limit reached).
Training time (sec)            : 0.6763

Settings
--------
Log-likelihood                 : 215.6778

Highest Positive Coefficients
-----------------------------
words[scribble]                : 19.1596
words[videozone]               : 17.8375
words[whirry]                  : 17.5098
words[nighteyes]               : 17.5098
words[tensionate]              : 17.3835

Lowest Negative Coefficients
----------------------------
words[newlwed]                 :

_3g) Evaluating the model using the testing data, from which the_ _**Testing accuracy**_ _comes out to be_ _**0.8696829855537721 (≈ 87.00%)**_.

In [7]:
model.evaluate(test)

{'accuracy': 0.8696829855537721,
 'auc': 0.9249718184658277,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  686  |
 |   positive   |     positive    |  4335 |
 |   positive   |     negative    |  613  |
 |   negative   |     negative    |  4334 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8696960577791153,
 'log_loss': 0.5506709434745038,
 'precision': 0.8633738299143597,
 'recall': 0.8761115602263541,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+---------------------+--------------------+------+------+
 | threshold |         fpr         |        tpr         |  p   |  n   |
 +-----------+---------------------+--------------------+------+------+
 |    0.0 

_3h) Creating a table of L1 and L2 penalties, based on which the models will be trained and the corresponing accuracies will be noted._

| | L1 penalty| L2 penalty |
| --- | --- | --- |
| 0 | 0.0 | 0.01 |
| 1 | 0.0 | 0.0 | 
| 2 | 0.1 | 0.0 |
| 3 | 0.1 | 0.01| 
| 4 | 0.0 | 0.001 |
| 5 | 1.0 | 1.0 |
| 6 | 0.0 | 1.0 |

In [8]:
model_1 = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty = 0.0, l2_penalty = 0.0)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92611

Number of coefficients      : 92612

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 0.087228     | 0.925769          | 0.849151            |

| 1         | 4        | 1.000000  | 0.206626     | 0.965816          | 0.869630            |

| 2         | 5        | 1.000000  | 0.284365     | 0.982961          | 0.888112            |

| 3         | 6        | 1.000000  | 0.359781     | 0.986484          | 0.889610            |

| 4         | 7        | 1.000000  | 0.438211     | 0.992637          | 0.897602            |

| 9         | 12       | 1.000000  | 0.842313     | 0.999921          | 0.877123            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [9]:
model_1

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 92612
Number of examples             : 38030
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 92611

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.0

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 10
Solver status                  : Completed (Iteration limit reached).
Training time (sec)            : 0.5103

Settings
--------
Log-likelihood                 : 167.4288

Highest Positive Coefficients
-----------------------------
words[sassier]                 : 35.211
words[wrangles]                : 35.211
words[scribble]                : 19.8448
words[anansie]                 : 17.0685
words[27x41]                   : 16.62

Lowest Negative Coefficients
----------------------------
words[galleghar]               : -16.

In [10]:
model_1.evaluate(test)

{'accuracy': 0.8678772070626003,
 'auc': 0.9206588158280387,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  677  |
 |   positive   |     positive    |  4308 |
 |   positive   |     negative    |  640  |
 |   negative   |     negative    |  4343 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8674116581093326,
 'log_loss': 0.640983340009139,
 'precision': 0.8641925777331996,
 'recall': 0.8706548100242523,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+---------------------+--------------------+------+------+
 | threshold |         fpr         |        tpr         |  p   |  n   |
 +-----------+---------------------+--------------------+------+------+
 |    0.0  

In [11]:
model_2 = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty = 0.1, l2_penalty = 0.0)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92715

Number of coefficients      : 92716

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.047946     | 0.500210          | 0.513986            |

| 1         | 2        | 0.000008  | 1.311521     | 0.925953          | 0.848152            |

| 2         | 3        | 0.000008  | 1.432297     | 0.956666          | 0.867632            |

| 3         | 4        | 0.000003  | 1.632728     | 0.970024          | 0.882118            |

| 4         | 5        | 0.000003  | 1.745757     | 0.973389          | 0.875624            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [12]:
model_2

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 92716
Number of examples             : 38030
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 92715

Hyperparameters
---------------
L1 penalty                     : 0.1
L2 penalty                     : 0.0

Training Summary
----------------
Solver                         : fista
Solver iterations              : 4
Solver status                  : SUCCESS: Optimal solution found.
Training time (sec)            : 1.783

Settings
--------
Log-likelihood                 : 8600.5424

Highest Positive Coefficients
-----------------------------
words[whirry]                  : 0.7565
words[nighteyes]               : 0.7565
words[videozone]               : 0.7505
words[midichlorians]           : 0.7299
words[conaughey]               : 0.7199

Lowest Negative Coefficients
----------------------------
words[milwaulkee]              : -0.7163
wo

In [13]:
model_2.evaluate(test)

{'accuracy': 0.8900481540930979,
 'auc': 0.9516329387381758,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  397  |
 |   positive   |     negative    |  699  |
 |   positive   |     positive    |  4249 |
 |   negative   |     negative    |  4623 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8857619345424222,
 'log_loss': 0.3637713803909529,
 'precision': 0.9145501506672407,
 'recall': 0.858730800323363,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+--------------------+-----+------+------+
 | threshold |        fpr         | tpr |  p   |  n   |
 +-----------+--------------------+-----+------+------+
 |    0.0    |        1.0         | 1.0 | 4948 | 5020 |
 | 

In [14]:
model_3 = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty = 0.1, l2_penalty = 0.01)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92689

Number of coefficients      : 92690

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.048121     | 0.501341          | 0.492507            |

| 1         | 2        | 0.000008  | 1.346919     | 0.926611          | 0.863636            |

| 2         | 3        | 0.000008  | 1.463019     | 0.954536          | 0.881618            |

| 3         | 4        | 0.000003  | 1.656770     | 0.969261          | 0.882118            |

| 4         | 5        | 0.000002  | 1.819782     | 0.974415          | 0.891608            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [15]:
model_3

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 92690
Number of examples             : 38030
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 92689

Hyperparameters
---------------
L1 penalty                     : 0.1
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : fista
Solver iterations              : 4
Solver status                  : SUCCESS: Optimal solution found.
Training time (sec)            : 1.8557

Settings
--------
Log-likelihood                 : 8604.2527

Highest Positive Coefficients
-----------------------------
words[whirry]                  : 0.7279
words[nighteyes]               : 0.7279
words[videozone]               : 0.7172
words[midichlorians]           : 0.7089
words[conaughey]               : 0.6951

Lowest Negative Coefficients
----------------------------
words[milwaulkee]              : -0.6888


In [16]:
model_3.evaluate(test)

{'accuracy': 0.8921548956661316,
 'auc': 0.9509631643192781,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  537  |
 |   positive   |     negative    |  538  |
 |   positive   |     positive    |  4410 |
 |   negative   |     negative    |  4483 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8913592723597777,
 'log_loss': 0.3659634052026425,
 'precision': 0.8914493632504549,
 'recall': 0.891269199676637,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+--------------------+-----+------+------+
 | threshold |        fpr         | tpr |  p   |  n   |
 +-----------+--------------------+-----+------+------+
 |    0.0    |        1.0         | 1.0 | 4948 | 5020 |
 | 

In [17]:
model_4 = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty = 0.0, l2_penalty = 0.001)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92722

Number of coefficients      : 92723

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 0.100763     | 0.924638          | 0.860639            |

| 1         | 4        | 1.000000  | 0.223625     | 0.960005          | 0.870629            |

| 2         | 6        | 1.000000  | 0.341523     | 0.982672          | 0.899600            |

| 3         | 7        | 1.000000  | 0.419503     | 0.986090          | 0.897103            |

| 4         | 9        | 1.000000  | 0.544587     | 0.994899          | 0.898102            |

| 9         | 14       | 1.000000  | 0.957489     | 0.999947          | 0.866134            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [18]:
model_4

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 92723
Number of examples             : 38030
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 92722

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.001

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 10
Solver status                  : Completed (Iteration limit reached).
Training time (sec)            : 0.6158

Settings
--------
Log-likelihood                 : 156.2673

Highest Positive Coefficients
-----------------------------
words[sebald]                  : 64.0062
words[rowers]                  : 53.1159
words[4w]                      : 53.1159
words[4m]                      : 53.1159
words[269]                     : 53.1159

Lowest Negative Coefficients
----------------------------
words[annotation]              

In [19]:
model_4.evaluate(test)

{'accuracy': 0.8633627608346709,
 'auc': 0.9162254780393366,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  632  |
 |   positive   |     positive    |  4218 |
 |   positive   |     negative    |  730  |
 |   negative   |     negative    |  4388 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8609920391916718,
 'log_loss': 0.666874771509709,
 'precision': 0.8696907216494846,
 'recall': 0.8524656426839127,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+---------------------+--------------------+------+------+
 | threshold |         fpr         |        tpr         |  p   |  n   |
 +-----------+---------------------+--------------------+------+------+
 |    0.0  

In [20]:
model_5 = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty = 1.0, l2_penalty = 1.0)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92691

Number of coefficients      : 92692

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.062505     | 0.501183          | 0.495504            |

| 1         | 2        | 0.000008  | 1.383256     | 0.926164          | 0.855644            |

| 2         | 3        | 0.000008  | 1.501335     | 0.952485          | 0.873626            |

| 3         | 4        | 0.000002  | 1.750219     | 0.970392          | 0.882617            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [21]:
model_5

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 92692
Number of examples             : 38030
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 92691

Hyperparameters
---------------
L1 penalty                     : 1.0
L2 penalty                     : 1.0

Training Summary
----------------
Solver                         : fista
Solver iterations              : 3
Solver status                  : SUCCESS: Optimal solution found.
Training time (sec)            : 1.7872

Settings
--------
Log-likelihood                 : 9756.4552

Highest Positive Coefficients
-----------------------------
words[midichlorians]           : 0.5584
words[whirry]                  : 0.5496
words[nighteyes]               : 0.5496
words[videozone]               : 0.5453
words[desensitizing]           : 0.5392

Lowest Negative Coefficients
----------------------------
words[echt]                    : -0.5428
w

In [22]:
model_5.evaluate(test)

{'accuracy': 0.8877407704654896,
 'auc': 0.9475743750946096,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  640  |
 |   positive   |     negative    |  479  |
 |   positive   |     positive    |  4469 |
 |   negative   |     negative    |  4380 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8887342149746444,
 'log_loss': 0.38630404598645984,
 'precision': 0.8747308670972793,
 'recall': 0.9031932093775262,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+--------------------+-----+------+------+
 | threshold |        fpr         | tpr |  p   |  n   |
 +-----------+--------------------+-----+------+------+
 |    0.0    |        1.0         | 1.0 | 4948 | 5020 |
 

In [23]:
model_6 = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty = 0.0, l2_penalty = 1.0)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92585

Number of coefficients      : 92586

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 0.106910     | 0.926874          | 0.854645            |

| 1         | 4        | 1.000000  | 0.233888     | 0.974625          | 0.882617            |

| 2         | 5        | 1.000000  | 0.312870     | 0.985853          | 0.880120            |

| 3         | 6        | 1.000000  | 0.400256     | 0.992190          | 0.889610            |

| 4         | 7        | 1.000000  | 0.486124     | 0.995162          | 0.897602            |

| 9         | 12       | 1.000000  | 0.911034     | 0.999606          | 0.866134            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [24]:
model_6

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 92586
Number of examples             : 38030
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 92585

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 1.0

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 10
Solver status                  : Completed (Iteration limit reached).
Training time (sec)            : 0.5492

Settings
--------
Log-likelihood                 : 193.5778

Highest Positive Coefficients
-----------------------------
words[scribble]                : 32.8148
words[videozone]               : 26.0495
words[whirry]                  : 25.8822
words[nighteyes]               : 25.8822
words[midichlorians]           : 24.8454

Lowest Negative Coefficients
----------------------------
words[hardiman]                : 

In [25]:
model_6.evaluate(test)

{'accuracy': 0.8687800963081862,
 'auc': 0.9233034313836003,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  680  |
 |   positive   |     positive    |  4320 |
 |   positive   |     negative    |  628  |
 |   negative   |     negative    |  4340 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8685162846803378,
 'log_loss': 0.5689002423422415,
 'precision': 0.864,
 'recall': 0.8730800323362975,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+---------------------+--------------------+------+------+
 | threshold |         fpr         |        tpr         |  p   |  n   |
 +-----------+---------------------+--------------------+------+------+
 |    0.0    |         

_3h) Recording the accuracies:_


| | L1 penalty| L2 penalty | Training accuracy | Validation accuracy | Testing accuracy | 
| --- | --- | --- | --- | --- | --- |
| 0 | 0.0 | 0.01 | 0.999448 | 0.868132 | 0.869683 |
| 1 | 0.0 | 0.0 | 0.999921 | 0.877123 | 0.867877 |
| 2 | 0.1 | 0.0 | 0.973389 | 0.875624 | 0.890048 |
| 3 | 0.1 | 0.01| 0.974415 | 0.891608 | 0.892155 |
| 4 | 0.0 | 0.001 | 0.999947 | 0.866134 | 0.863363 |
| 5 | 1.0 | 1.0 | 0.970392 | 0.882617 | 0.887741 | 
| 6 | 0.0 | 1.0 | 0.999606 | 0.866134 | 0.868780 | 

_3i) The optimal choice of L1 and L2 penalties from the above recording would be_ _**(3) 0.1 and 0.01**_ _respectively. This is because the all three accuracies are fairly high. With high Training accuracy, the validation and testing accuracies are also relatively good for this regularisation. Even the number of epochs used to reach this accuracy was less as compared to most of the other models._   